### PR TITLE
Build neural net blocks

### PR DESCRIPTION
**Pull request by janik072**
Added 3 math function (softmax, swiglu, rmsnorm) including several tests for each of them.

### CODE REVIEW

File: src/math_llama.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""
+    math_llama.jl
+
+    This file contains the math functions used in the Llama model.
+"""
+
+begin
```
Comment (kleincode) [inline src/math_llama.jl:7]: The outer `begin ... end` is not needed

File: src/math_llama.jl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# Calculate the root mean square norm of a vector. Reference in llama2.c lines 182-195
```
Comment (kleincode) [inline src/math_llama.jl:8]: Please use proper docstrings (with `"""`, the function name indented by one tab), see https://docs.julialang.org/en/v1/manual/documentation/#Writing-Documentation or code on the current main branch
Add a docstring for every function and struct, so it will be added automatically to the documentation.

File: src/math_llama.jl [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+function rmsnorm(x::Vector{Float32}, weight::Vector{Float32})
+    size = length(x)
+    # Calculate the sum of the squares
+    sum_squares = sum(x .^ 2) / size
```
Comment (kleincode) [inline src/math_llama.jl:12]: Inline size as it is only used once

File: src/math_llama.jl [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    sum_squares += 1e-5
```
Comment (kleincode) [inline src/math_llama.jl:13]: Would probably be better to stay in the Float32 domain, i.e. `1f-5` instead of `1e-5` and `1.0f0` instead of `1.0`. As the rmsnorm function is called pretty often, this might make a small difference :D

File: src/math_llama.jl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    sum_squares = 1.0 / sqrt(sum_squares) 
+    
+    o = weight .* (sum_squares .* x)
+    
+    return o
```
Comment (kleincode) [inline src/math_llama.jl:18]: To make the code more readable, I would suggest to remove some whitespaces and maybe return the expression assigned to `o` directly (as `o` is only used once)

File: src/math_llama.jl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+# Calculate the softmax of a vector. Reference in llama2.c lines 197-215
+function softmax(x::Vector{Float32})::Vector{Float32}
+
+    # exp and sum
+    sum_exp = exp.(x .- maximum(x))
+
+    # normalize
+    res = sum_exp ./ sum(sum_exp)
+
+    return res
```
Comment (kleincode) [inline src/math_llama.jl:31]: Again, return the expression assigned to `res` directly

File: test/test_math_llama.jl [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        res_4 = rmsnorm(x, weight)
+        exp_4 = [0.00031622764, 0.00063245529, 0.00189736588]
+        @test isapprox(res_4, exp_4, atol=1e-6)
+
+        # Test 5: Edge case with single element
+        x = Float32[3.0]
+        weight = Float32[2.0]
+        res_5 = rmsnorm(x, weight)
+        exp_5 = [1.999998888889815]
+        @test res_5 ≈ exp_5
```
Comment (kleincode) [inline test/test_math_llama.jl:39]: Optional test case: input empty list (should also return empty list)
(also applies to the other two functions)
In general, it is nice to have predictable behavior - either your function throws or it returns an empty list (probably the better option). Either way, you should know what will happen then.

**Review by kleincode — CHANGES_REQUESTED** [review]
Please add docstrings, some code style suggestions. Otherwise, looks great!
When you're finished, please merge main and re-request review.

**Review by kleincode — APPROVED** [review]
